src/api/iterator-walker.ts

```diff
@@ -0,0 +1,285 @@
+import { basename, dirname } from "path";
+import { isRootDirectory, normalizePath } from "../utils";
+import { WalkerState, Options, OutputIterator, IterableOutput } from "../types";
+import * as joinPath from "./functions/join-path";
+import * as resolveSymlink from "./functions/resolve-symlink";
+import { Queue } from "./queue";
+import type { Dirent } from "fs";
+import * as nativeFs from "fs";
+import { Counter } from "./counter";
+import { Aborter } from "./aborter";
+import { promisify } from "node:util";
+
+export class IteratorWalker<TOutput extends IterableOutput> {
+  private readonly root: string;
+  private readonly state: WalkerState;
+  private readonly joinPath: joinPath.JoinPathFunction;
+  private readonly resolveSymlink: resolveSymlink.ResolveSymlinkFunction | null;
+
+  constructor(root: string, options: Options) {
+    this.root = normalizePath(root, options);
+    this.state = {
+      root: isRootDirectory(this.root) ? this.root : this.root.slice(0, -1),
+      paths: [],
+      groups: [],
+      counts: new Counter(),
+      options,
+      queue: new Queue(() => {}),
+      symlinks: new Map(),
+      visited: [""].slice(0, 0),
+      controller: new Aborter(),
+      fs: options.fs || nativeFs,
+    };
+
+    this.joinPath = joinPath.build(this.root, options);
+    this.resolveSymlink = resolveSymlink.build(options, false);
+  }
+
+  get aborted(): boolean {
+    const {
+      controller,
+      options: { signal },
+    } = this.state;
+    return controller.aborted || (signal !== undefined && signal.aborted);
+  }
+
+  #shouldPushDirectory(directoryPath: string): boolean {
+    const { options } = this.state;
+    const { includeDirs, filters } = options;
+
+    if (!includeDirs) {
+      return false;
+    }
+
+    if (filters && filters.length) {
+      return filters.every((filter) => filter(directoryPath, true));
+    }
+
+    return true;
+  }
+
+  #normalizeDirectoryPath(path: string): string {
```

Comment:
this breaks node 12, even with tsdown. should be an arrow function 👀

Reply:
afaict its not that it should be an arrow function, its that im using private members. we can just name it like the others (using typescript `private`, without `#`)